tracer/src/Datadog.Trace/Util/UnmanagedMemoryPool.cs

```diff
@@ -158,4 +191,46 @@ private void ThrowObjectDisposedException()
     {
         throw new ObjectDisposedException("UnmanagedMemoryPool");
     }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private void OnPoolCreated()
+    {
+        if (_fastPoolCount < WafConstants.MaxUnmanagedPools)
+        {
+            Interlocked.Increment(ref _fastPoolCount);
+            Log.Debug<int, int>("Created fast WAF unmanaged pool. Current pools -> Fast: {PoolCount}  Slow: {SlowPoolCount}", _fastPoolCount, _slowPoolCount);
+            TelemetryFactory.Metrics.RecordGaugePoolCount(_fastPoolCount);
+            _isSlow = false;
+        }
+        else
+        {
+            Interlocked.Increment(ref _slowPoolCount);
+            Log.Debug<int, int>("Created slow WAF unmanaged pool. Current pools -> Fast: {PoolCount}  Slow: {SlowPoolCount}", _fastPoolCount, _slowPoolCount);
+            TelemetryFactory.Metrics.RecordGaugePoolSlowCount(_slowPoolCount);
+            _isSlow = true;
+        }
+
+        RegisterTheoreticalMaxMemUsage();
+    }
+
+    private void OnPoolDestroyed()
+    {
+        if (_isSlow)
+        {
+            Interlocked.Decrement(ref _slowPoolCount);
+        }
+        else
+        {
+            Interlocked.Decrement(ref _fastPoolCount);
+        }
+
+        RegisterTheoreticalMaxMemUsage();
+    }
+
+    private void RegisterTheoreticalMaxMemUsage()
+    {
+        // Calculate theoretical max memory consumed by this pool
+        var maxMem = (_blockSize + sizeof(IntPtr)) * _length;
```

Comment:
this could be calculated only once at pool constructor as all variables here are readonly?